### PR TITLE
Defined API for initialization of cgroups

### PR DESCRIPTION
**Pull request by narrieta**
## Description

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1530)

### CODE REVIEW

File: azurelinuxagent/common/cgroupapi.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+                        raise CGroupsException("Create directory for cgroup {0}: normal file already exists with that name".format(path))
+                    else:
+                        pass # There was a race to create the directory, but it's there now, and that's fine
+                elif e.errno == errno.EACCES:
+                    # This is unexpected, as the agent runs as root
+                    raise CGroupsException("Create directory for cgroup {0}: permission denied".format(path))
+                else:
+                    raise
+
+    def _foreach_controller(self, operation, message):
```
Comment (pgombar) [inline azurelinuxagent/common/cgroupapi.py]: This method can be static.

File: azurelinuxagent/common/cgroupconfigurator.py [hunk 2]
```diff
@@ -107,32 +102,36 @@ def __init__(self, name, path_maker, limits=None):
...
+    def is_systemd():
         """
-        Determine if systemd is managing system services. Many extensions are structured as a set of services,
-        including the agent itself; systemd expects those services to remain in the cgroups in which it placed them.
-        If this process (presumed to be the agent) is in a cgroup that looks like one created by systemd, we can
-        assume systemd is in use.
+        Determine if systemd is managing system services. If this process (presumed to be the agent) is in a CPU cgroup
+        that looks like one created by systemd, we can assume systemd is in use.
+
+        NOTE: This function has the side effect of mounting the cgroups file system if it is not already mounted.
```
Comment (pgombar) [inline azurelinuxagent/common/cgroupconfigurator.py]: When does this happen? I don't see where the mounting happens here.

File: azurelinuxagent/common/cgroupconfigurator.py [hunk 3]
```diff
@@ -511,3 +431,73 @@ def get_default_memory_limits(cgroup_name):
...
+
+            add_event(
+                AGENT_NAME,
+                version=CURRENT_VERSION,
+                op=WALAEventOperation.InitializeCGroups,
+                is_success=CGroupConfigurator.enabled(),
+                message=status,
+                log_event=False)
+
+        def _invoke_cgroup_operation(self, operation, error_message):
```
Comment (pgombar) [inline azurelinuxagent/common/cgroupconfigurator.py]: This can be static.

File: azurelinuxagent/common/cgroupapi.py (hunk 1 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py]: yes, fixed

File: azurelinuxagent/common/cgroupconfigurator.py (hunk 3 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroupconfigurator.py]: thanks, fixed

File: azurelinuxagent/common/cgroupconfigurator.py [hunk 4]
```diff
@@ -511,3 +431,73 @@ def get_default_memory_limits(cgroup_name):
...
+            """
+            Ensures the given operation is invoked only if cgroups are enabled and traps any errors on the operation
+            """
+            if not CGroupConfigurator.enabled():
+                return
+
+            try:
+                operation()
```
Comment (vrdmr) [inline azurelinuxagent/common/cgroupconfigurator.py]: What does this do? Or is this a stub for now?

File: azurelinuxagent/common/cgroupconfigurator.py (hunk 2 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroupconfigurator.py]: comment was for a previous iteration of the code... i'll remove it

**Comment by narrieta** [comment]
@pgombar thanks for your comments. I updated the code, could you take a look at the last commit?

File: azurelinuxagent/common/cgroupconfigurator.py (hunk 4 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroupconfigurator.py]: it's the call to the operation passed as argument... lines 488 & 494 are examples of its use

Comment (vrdmr) [reply azurelinuxagent/common/cgroupconfigurator.py]: Got it.

**Pull request merged**